test/integration/federation/framework/crud.go

```diff
@@ -0,0 +1,199 @@
+	clusterClients []clientset.Interface
+	waitInterval   time.Duration
+	// Federation operations will use wait.ForeverTestTimeout.  Any
+	// operation that involves member clusters may take longer due to
+	// propagation latency.
+	clusterWaitTimeout time.Duration
+}
+
+func NewCRUDHelperWithTimeout(logger TestLogger, adapter ResourceAdapter, clusterClients []clientset.Interface, waitInterval, clusterWaitTimeout time.Duration) *CRUDHelper {
+	return &CRUDHelper{
+		t:                  logger,
+		adapter:            adapter,
+		kind:               adapter.Kind(),
+		clusterClients:     clusterClients,
+		waitInterval:       waitInterval,
+		clusterWaitTimeout: clusterWaitTimeout,
+	}
+}
+
+func NewCRUDHelper(t *testing.T, adapter ResourceAdapter, clusterClients []clientset.Interface) *CRUDHelper {
+	logger := &IntegrationLogger{t}
+	return NewCRUDHelperWithTimeout(logger, adapter, clusterClients, DefaultWaitInterval, wait.ForeverTestTimeout)
+}
+
+func (c *CRUDHelper) CheckLifecycle(obj pkgruntime.Object) {
+	updatedObj := c.CheckCreate(obj)
+	c.CheckUpdate(updatedObj)
+
+	// Validate the golden path - removal of dependents
+	orphanDependents := false
+	c.CheckDelete(updatedObj, &orphanDependents)
+}
+
+func (c *CRUDHelper) CheckCreate(obj pkgruntime.Object) pkgruntime.Object {
+	nsName := c.adapter.NamespacedName(obj)
+
+	c.t.Logf("Creating federated %s %q", c.kind, nsName)
+
+	obj, err := c.adapter.FedCreate(obj)
+	if err != nil {
+		c.t.Fatalf("Error creating federated %s %q : %v", c.kind, nsName, err)
+	}
+
+	c.checkPropagation(obj)
+
+	return obj
+}
+
+func (c *CRUDHelper) CheckUpdate(obj pkgruntime.Object) {
+	nsName := c.adapter.NamespacedName(obj)
+
+	// Insure that the provided object does not have labels, which
+	// would break the simple check for mutation.
+	meta := c.adapter.ObjectMeta(obj)
+	if len(meta.Labels) != 0 {
+		c.t.Fatalf("Expected federated %s %q to have been created without labels", c.kind, nsName)
+	}
+
+	c.t.Logf("Updating federated %s %q", c.kind, nsName)
+	updatedObj, err := c.updateFedObject(obj)
+	if err != nil {
+		c.t.Fatalf("Error updating federated %s %q: %v", c.kind, nsName, err)
+	}
+
+	// updateFedObject is expected to have added a label
+	meta = c.adapter.ObjectMeta(updatedObj)
+	if len(meta.Labels) == 0 {
+		c.t.Fatalf("Federated %s %q not mutated", c.kind, nsName)
+	}
+
+	c.checkPropagation(updatedObj)
+}
+
+func (c *CRUDHelper) CheckDelete(obj pkgruntime.Object, orphanDependents *bool) {
+	nsName := c.adapter.NamespacedName(obj)
+
+	c.t.Logf("Deleting federated %s %q", c.kind, nsName)
+	err := c.adapter.FedDelete(nsName, &metav1.DeleteOptions{OrphanDependents: orphanDependents})
+	if err != nil {
+		c.t.Fatalf("Error deleting federated %s %q: %v", c.kind, nsName, err)
+	}
+
+	// Wait for deletion.  The federation resource will only be removed once orphan deletion has been
+	// completed or deemed unnecessary.
+	err = wait.PollImmediate(c.waitInterval, c.clusterWaitTimeout, func() (bool, error) {
```

Comment:
I'm using the potentially longer cluster timeout because deletion may depend on removal of the resource from the member clusters.  The federation timeout is used where an operation is only targeting the federation api.  I'll add a comment to make that clear.